lib/client/db/mcp/managedconn.go

```diff
@@ -0,0 +1,164 @@
+package mcp
+
+import (
+	"context"
+	"log/slog"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+)
+
+// ManagedConn manages a database connection, providing a way to keep a single
+// active connection per database, and also closing the connection once it
+// becomes inactive.
+//
+// Managed connections are always used with Exec function.
+type ManagedConn[T any, C conn[T]] struct {
+	mu   sync.Mutex
+	cond sync.Cond
+
+	cfg      *ManagedConnConfig
+	newFunc  func(context.Context) (*T, error)
+	active   C
+	closed   bool
+	watchdog *time.Timer
+
+	// cancelExec is the context cancelation function of a execution. It will
+	// only be set when there is a running execution, so it can be used to
+	// indicate the connection is in use.
+	cancelExec context.CancelFunc
+}
+
+// ManagedConnConfig represents a managed connection config.
+type ManagedConnConfig struct {
+	// MaxIdleTime is the max connection idle time before it gets closed
+	MaxIdleTime time.Duration
+	// Logger is the logger used by the managed connection.
+	Logger *slog.Logger
+}
+
+// CheckAndSetDefaults checks and sets the defaults
+func (c *ManagedConnConfig) CheckAndSetDefaults() error {
+	if c.MaxIdleTime == 0 {
+		return trace.BadParameter("max idle time must be greater than zero")
+	}
+	if c.Logger == nil {
+		c.Logger = slog.Default()
+	}
+	return nil
+}
+
+// conn represents the database connection.
+type conn[T any] interface {
+	*T
+	// Close closes the connection.
+	Close(context.Context) error
+	// IsClosed returns true if the connection is closed.
+	IsClosed() bool
+}
```

Comment:
I think we can make `conn` non-generic and express the constraint on T to be a pointer instead. That way we might be able to drop the number of params to just one.